usl_pipeline/cloud_functions/main.py

```diff
@@ -862,3 +869,80 @@ def _write_as_npy(blob: storage.Blob, array: NDArray) -> None:
     npy_file.flush()
     npy_file.seek(0)
     blob.upload_from_file(npy_file)
+
+
+def _start_feature_rescaling_if_ready(feature_bucket: storage.Bucket, blob_path: str):
+    """Creates scaled version for every chunk file in the storage bucket parent folder.
+
+    Output files with scaled feature matrices will be placed to the same folder where
+    unscaled files (following "chunk_*" name pattern) are stored but will get "scaled_"
+    prefix. Rescaling only starts when metadata of study area (corresponding to the
+    parent folder name) contains chunks counters giving exactly the same number of
+    chunks as are currently stored in the bucket folder. This should happen when the
+    last feature chunk is stored.
+
+    Args:
+        feature_bucket: Storage bucket with feature matrices.
+        blob_path: Path pointing to one of unscaled feature matrix files with the name
+            following the "chunk_*" pattern. Other files are ignored.
+    """
+    feature_path = pathlib.PurePosixPath(blob_path)
+    if not feature_path.name.startswith("chunk_"):
+        return
+    feature_parent_path = feature_path.parent
+    study_area_name = feature_parent_path.parts[-1]
+
+    db = firestore.Client()
+    study_area = metastore.StudyArea.get(db, study_area_name)
+    if study_area.chunk_x_count is None or study_area.chunk_y_count is None:
+        logging.info(
+            f"[Feature Rescaler] Study area {study_area_name} has no chunk counts in "
+            + "metadata"
+        )
+        return
+
+    feature_blobs = list(
+        feature_bucket.list_blobs(prefix=f"{feature_parent_path}/chunk_")
+    )
+
+    chunk_count = study_area.chunk_x_count * study_area.chunk_y_count
+    if chunk_count != len(feature_blobs):
+        logging.info(
+            f"[Feature Rescaler] Study area {study_area_name} has chunk counts"
+            + f" in metadata ({chunk_count}) different from number of stored feature"
+            + f" chunks ({len(feature_blobs)})"
+        )
+        return
+
+    # We're at the final step when all unscaled feature matrices are ready
+    for feature_blob in feature_blobs:
+        logging.info(f"[Feature Rescaler] Rescaling feature matrix {feature_blob.path}")
+        output_blob = feature_bucket.blob(
+            f"{feature_parent_path}/scaled_{feature_blob.name}"
+        )
+        with feature_blob.open("rb") as in_fd, output_blob.open("wb") as out_fd:
+            feature_raster_transformers.rescale_feature_blob(in_fd, study_area, out_fd)
+    # Once all chunks are processed we can delete unscaled ones:
+    for feature_blob in feature_blobs:
+        feature_blob.delete()
+
+
+@functions_framework.cloud_event
+@_retry_and_report_errors()
+def rescale_feature_matrices(cloud_event: functions_framework.CloudEvent) -> None:
```

Comment:
What were you envisioning would trigger this function?
If the unscaled outputs are written to the features bucket, it will trigger the `build_feature_matrix` function again. Triggers run for all uploads to anywhere in the bucket. You can't have a trigger on certain paths. 
For the collapsing the labels, I rely on the fact that step 1 has .rsl files as input and step 2 has .npy files as input, so there's one cloud function, but it dispatches based on the file type
https://github.com/UrbanSystemsLab/climateiq-cnn/blob/ed43aa9375d33d33d65d71a7c47d172075e72092/usl_pipeline/cloud_functions/main.py#L372

I think we'd do similar here -- have one function that first checks to see if the file path is for a scaled or unscaled file and dispatch based on that

Reply:
Walt, I think you're mixing up buckets. `build_feature_matrix` CF is triggered when new file is stored to `study-area-chunks` bucket. This new CF function (`rescale_feature_matrices`) is going to be attached to `study-area-feature-chunks` bucket which doesn't currently have any attached triggers (as far as I know). Please correct me if I'm wrong.

Reply:
Ah! Okay, that's why I was asking -- I wasn't sure if you were planning to trigger this function off of study-area-chunks or study-area-feature-chunks writes.
Yes, you're correct that that we don't have any triggers for feature-chunks yet.

Reply:
We will want to avoid applying flood logic (like this scaling) to heat files written to the study-area-feature-chunks bucket as well.

Reply:
I think I didn't change current WRF logic by my PR. So I guess let's separate refactoring effort decoupling the logic between Flood and WRF into separate PR? I don't want to change WRF logic in this PR.

Reply:
Yep, `needs_scaling` field was added.